src/runtime_src/core/common/api/xrt_module.cpp

```diff
@@ -905,24 +942,41 @@ class module_sram : public module_impl
     XRT_PRINTF("-> module_sram::create_instr_buf()\n");
     const auto& data = parent->get_instr();
     size_t sz = data.size();
-
-    if (sz == 0) {
-      std::cout << "instr buf is empty" << std::endl;
-      return;
-    }
+    if (sz == 0)
+      throw std::runtime_error("Invalid instruction buffer size");
 
     // create bo combined size of all ctrlcodes
     m_instr_bo = xrt::bo{ m_hwctx, sz, xrt::bo::flags::cacheable, 1 /* fix me */ };
 
     // copy instruction into bo
-    fill_instr_buf(m_instr_bo, data);
+    fill_bo_with_data(m_instr_bo, data);
 
 #ifdef _DEBUG
     dump_bo(m_instr_bo, "instrBo.bin");
 #endif
 
+    const auto& preempt_save_data = parent->get_preempt_save();
+    auto preempt_save_data_size = preempt_save_data.size();
+    if (preempt_save_data_size > 0) {
+      m_preempt_save_bo = xrt::bo{ m_hwctx, preempt_save_data_size, xrt::bo::flags::cacheable, 1 /* fix me */ };
+      fill_bo_with_data(m_preempt_save_bo, preempt_save_data);
+    }
+
+    const auto& preempt_restore_data = parent->get_preempt_restore();
+    auto preempt_restore_data_size = preempt_restore_data.size();
+    if (preempt_restore_data_size > 0) {
+      m_preempt_restore_bo = xrt::bo{ m_hwctx, preempt_restore_data_size, xrt::bo::flags::cacheable, 1 /* fix me */ };
+      fill_bo_with_data(m_preempt_restore_bo, preempt_restore_data);
+    }
+
+    if ((preempt_save_data_size > 0) && (preempt_restore_data_size > 0)) {
+      m_scratchmem = xrt::ext::bo{ m_hwctx, m_scratchmem_size };
```

Comment:
how are we handling sync_bo of this buffer?

Reply:
Interface function get_scratchmem is added, once the result come back to this bo, use the function to sync bo.